pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,220 @@
+        return circuits, sum_fn
+
+    # Expand each of the broadcasted sum-expanded circuits
+    expanded_tapes, expanded_fn = qml.transforms.map_batch_transform(
+        qml.transforms.broadcast_expand, circuits
+    )
+
+    # Chain the postprocessing functions of the broadcasted-tape expansions and the Sum
+    # expansion. Note that the application order is reversed compared to the expansion order,
+    # i.e. while we first applied `sum_expand` to the tape, we need to process the
+    # results from the broadcast expansion first.
+    def total_processing(results):
+        return sum_fn(expanded_fn(results))
+
+    return expanded_tapes, total_processing
+
+
+def check_validity(tape):
+    """Checks whether the operations and observables in queue are all supported by the device.
+
+    Args:
+        tape (.QuantumTape): tape from which to validate operations and observables
+
+    Raises:
+        DeviceError: if there are operations in the queue or observables that the device does
+            not support
+    """
+    for o in tape.operations:
+        operation_name = o.name
+
+        if not _stopping_condition(o):
+            raise DeviceError(f"Gate {operation_name} not supported on Python Device")
+
+    for o in tape.observables:
+        if isinstance(o, Tensor):
+            for i in o.obs:
+                if not _supports_observable(i.name):
+                    raise DeviceError(f"Observable {i.name} not supported on Python Device")
+        else:
+            observable_name = o.name
+
+            if not _supports_observable(observable_name):
+                raise DeviceError(f"Observable {observable_name} not supported on Python Device")
+
+
+def preprocess(tapes, execution_config=None, max_expansion=10):
+    """Preprocess a batch of `QuantumTape` objects to make them ready for execution.
+
+    This function validates a batch of `QuantumTape` objects by transforming and expanding
+    them to ensure all operators and measurements are supported by the execution device.
+
+    Args:
+        tapes (Sequence[QuantumTape]): Batch of tapes to be processed.
+        execution_config (.ExecutionConfig): execution configuration with configurable
+            options for the execution.
+        max_expansion (int): The number of times the circuit should be
+            expanded. Expansion occurs when an operation or measurement is not
+            supported, and results in a gate decomposition. If any operations
+            in the decomposition remain unsupported by the device, another
+            expansion occurs.
+
+    Returns:
+        Tuple[Sequence[.QuantumTape], callable]: Returns a tuple containing
+        the sequence of circuits to be executed, and a post-processing function
+        to be applied to the list of evaluated circuit results.
+    """
+    # Finite shot support will be added later
+    if execution_config and getattr(execution_config, "shots", None) is not None:
```

Comment:
if it exists (which you confirm), we can assume it has a `shots` attribute
```suggestion
    if execution_config and execution_config.shots is not None:
```

Reply:
Also, I created a global `DefaultExecutionConfig` object in the divine interface PR which I use as the default value in the call signature, instead of using `None`.